pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,261 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package trace
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgmodel/metrics"
+	"github.com/timescale/promscale/pkg/tracer"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	defaultReqBufferSize      = 100000                 // buffer for incoming requests, especially important for async acks
+	DefaultBatchSize          = 1000                   // this is soft limit as we might produce bigger batches in some cases
+	DefaultBatchTimeout       = 100 * time.Millisecond // we should aways aim at reching size limits, not timeout
+	defaultMaxBufferedBatches = 2000                   //  We want to avoid wait on batches
+)
+
+var defaultBatchWorkers int
+
+func init() {
+	defaultBatchWorkers = runtime.NumCPU() / 2
+	if defaultBatchWorkers == 0 {
+		// in case there <=1 CPU
+		defaultBatchWorkers = 1
+	}
+}
+
+// We batch individual insertTracesReq
+type Batch struct {
+	traces    ptrace.Traces
+	spanCount int
```

Comment:
I wonder if `spanCount` is needed. We are using it for `tb.spanCount >= tb.maxSize` which we can get by `tb.traces.SpanCount() >= tb.maxSize`.

Reply:
It's there to avoid recomputations since `traces.SpanCount()` has a nested `for` loops.